### ISSUE TITLE
Ablation study butterfly

### ISSUE THREAD

**Issue by thodkatz** [thodkatz/thesis#15]
I tried to understand the important components of butterfly model, and why it has such a good performance.

```python
      # """ train for discriminator """
      loss_d = self.forward_discriminator(
          batch_samples, RNA_input_dim, ATAC_input_dim, d_loss, "train"
      )
      self.optimizer_discriminator_R.zero_grad()
      self.optimizer_discriminator_A.zero_grad()
      loss_d.backward()
      self.optimizer_discriminator_R.step()
      self.optimizer_discriminator_A.step()
      
    
      """ train for generator """
      weight_temp = loss_weight.copy()
      if epoch < translation_kl_warmup:
          weight_temp[5] = loss_weight[5] * epoch / translation_kl_warmup
                          
      # self.forward_discriminator(
      #     batch_samples, RNA_input_dim, ATAC_input_dim, d_loss, "test"
      # )
                          
      reconstruct_loss, kl_div, loss_g = self.forward_translator(
          batch_samples,
          RNA_input_dim,
          ATAC_input_dim,
          a_loss,
          r_loss,
          weight_temp,
          "train",
          kl_mean,
      )
      
      #loss_d = torch.tensor(0)
    
      # if loss_d.item() < 1.35:
      #     loss_g -= loss_weight[2] * loss_d
      #     #print(f"loss_g: {loss_g.item()}, loss_weight: {loss_weight[2]}, loss_d: {loss_d.item()}")
      # else:
      #     pass
      #     #print(f"loss_g: {loss_g.item()}")
    
      self.optimizer_translator.zero_grad()
      if not lock_encoder_and_decoder:
          self.optimizer_R_encoder.zero_grad()
          self.optimizer_A_encoder.zero_grad()
          self.optimizer_R_decoder.zero_grad()
          self.optimizer_A_decoder.zero_grad()
      loss_g.backward()
      self.optimizer_translator.step()
      if not lock_encoder_and_decoder:
          self.optimizer_R_encoder.step()
          self.optimizer_A_encoder.step()
          self.optimizer_R_decoder.step()
          self.optimizer_A_decoder.step()
```

I was trying to understand the effect of the discriminator. I run some experiments, that either don't train the discriminator, or the loss_d isn't utilized to update the parameters of the generator.

- The loss_g -= factor * loss_d, effectively isn't utilized for the Nault dataset, so the discriminator loss didn't have impact to the generator one. But still as it seems the discriminator was affecting the performance.
- One reason was the batch norm updates in the encoder, due to the forward pass of the discriminator, irrelevant if any losses are computed for the encoder side.
- I noticed that when I use .detach() to the encoders of RNA, and ATAC, the discriminator even if it is trained or not, have zero influence to the network. But although the loss_d isn't utilized, when no using .detach(), somehow the encoders have been influenced.
- I am not sure how they are affected, since no optimizer or loss are associated with the discriminator and the encoder's parameters. How the gradient graph is used in this case? Since we do .backward(), shouldn't .grad of the encoder be lost, and parameters not updated, since we optimized only discriminator's parameters?
